jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java

```diff
@@ -469,66 +466,130 @@ protected void sendWelcome(HttpContent content, String pathInContext, boolean en
     }
 
-    protected void handleWelcomeAction(Request request, Response response, Callback callback, WelcomeAction welcomeAction) throws IOException
+    protected boolean handleWelcomeAction(Request request, Response response, Callback callback, WelcomeAction welcomeAction) throws Exception
     {
-        switch (welcomeAction.type)
+        return switch (welcomeAction.mode)
         {
-            case REDIRECT ->
-            {
-                response.getHeaders().putLongField(HttpHeader.CONTENT_LENGTH, 0);
-                sendRedirect(request, response, callback, welcomeAction.target);
-            }
+            case REDIRECT -> redirectWelcome(request, response, callback, welcomeAction.target);
             case SERVE ->
-            {
                 // TODO : check conditional headers.
-                HttpContent c = _contentFactory.getContent(welcomeAction.target);
-                sendData(request, response, callback, c, List.of());
-            }
-        }
+                serveWelcome(request, response, callback, welcomeAction.target);
+            case REHANDLE -> rehandleWelcome(request, response, callback, welcomeAction.target);
+        };
     }
 
-    private WelcomeAction processWelcome(Request request, Response response) throws IOException
+    /**
+     * <p>Redirects to the given welcome target.</p>
+     * <p>Implementations should use HTTP redirect APIs to generate
+     * a redirect response whose location is the welcome target.</p>
+     *
+     * @param request the request to redirect
+     * @param response the response
+     * @param callback the callback to complete
+     * @param welcomeTarget the welcome target to redirect to
+     * @return true when the request will be redirected and the callback completed,
+     * false otherwise
+     * @throws Exception if the redirection fails
+     */
+    protected boolean redirectWelcome(Request request, Response response, Callback callback, String welcomeTarget) throws Exception
     {
-        String welcomeTarget = _welcomeFactory.getWelcomeTarget(request);
+        response.getHeaders().putLongField(HttpHeader.CONTENT_LENGTH, 0);
+        sendRedirect(request, response, callback, welcomeTarget);
+        return true;
+    }
+
+    /**
+     * <p>Serves the given welcome target.</p>
+     * <p>Implementations should write the welcome
+     * target bytes over the network to the client.</p>
+     *
+     * @param request the request
+     * @param response the response
+     * @param callback the callback to complete
+     * @param welcomeTarget the welcome target to serve
+     * @return true when the welcome target will be served and the callback completed,
+     * false otherwise
+     * @throws Exception if serving the welcome target fails
+     */
+    protected boolean serveWelcome(Request request, Response response, Callback callback, String welcomeTarget) throws Exception
+    {
```

Comment:
Should this method check the conditional headers as per the TODO above?

Reply:
Perhaps, but in another PR.

Reply:
Can you open an issue for that.  Serving /index.html is a very frequent target and will often have conditional headers.

Reply:
#9632.